core/objects.go

```diff
@@ -146,6 +147,38 @@ type ValidationRecord struct {
 	AddressesTried []net.IP `json:"addressesTried,omitempty"`
 }
 
+type RehydratedValidationRecordFields struct {
+	Hostname string
+	Port     string
+}
+
+// TransformURLIntoHostnameAndPort takes a URL string as input and returns a
+// *RehydratedValidationRecordFields. To save database space, boulder no longer
+// stores the hostname and port fields in the validation record for an HTTP-01
+// challenge because that data can be rehydrated from the url field.
+func TransformURLIntoHostnameAndPort(input string) (*RehydratedValidationRecordFields, error) {
+	url, err := url.Parse(input)
+	if err != nil {
+		return &RehydratedValidationRecordFields{}, fmt.Errorf("could not parse url %v", input)
+	}
+
+	// A safety check for the scheme being either https or http is done in the
+	// VA prior to entering this transformation.
+	scheme := url.Scheme
+	hostname := url.Hostname()
+	if hostname == "" {
+		return &RehydratedValidationRecordFields{}, fmt.Errorf("could not parse hostname: %v", input)
+	}
```

Comment:
1. No need to allocate hostname to another variable, we can just used `url.Hostname()`.
2. Lets leave scheme as is until we're past the hostname check.
3. If we get an error in the logs from this it would be nice to know which URL resulted in an empty hostname.

```suggestion
	if url.Hostname() == "" {
		return "", "", fmt.Errorf("could not parse hostname from %q", vr.URL)
	}
```